### PR TITLE
Revised 4-pass algorithm, clearer explanation

### LINKED ISSUE #196: The Feistel network would benefit from a picture
A lot.

### PR DESCRIPTION
**Pull request by martinduke**
Attempt to fix #196, #197, #200.

Examples and test vectors not yet updated. PTAL at the description.

- Is the algorithm accurately described
- Are things clearer or not

### CODE REVIEW

**Review by huitema — APPROVED** [review]
Yes, the new text seems clear, and correct.

File: draft-ietf-quic-load-balancers.md [hunk 1]
```diff
@@ -436,91 +436,180 @@ for decryption. To understand this algorithm, it is useful to define four
...
+cipher to obtain a ciphertext.
+
+4. XOR the most significant bits of the ciphertext with right_0 to form
+right_1. If plaintext_len is odd, clear the four most significant bits. Steps 3
+and 4 can be summarized as
+  
+```
+   right_1 = XOR(right_0,
+                truncate(AES_ECB(key, expand(plaintext_len, 1, left_0)),
+                         half_len)
```
Comment (martinthomson) [inline draft-ietf-quic-load-balancers.md:563]: Assigning `truncate(AES_ECB(key, expand(plaintext_len, 1, left_0))` to an intermediate variable might make this fit line lengths easier.  If you don't, then this is poorly aligned with the line above.

This (and the later ones) appear to be missing a closing paren

**Review by martinthomson — COMMENTED** [review]
Much clearer, thanks.

I am concerned that you aren't being very clear about whether you are logically treating the intermediate values as bitstrings or integers and that could end up being confusing.  If not now, when people start implementing this and talking to each other about it.

File: draft-ietf-quic-load-balancers.md [hunk 2]
```diff
@@ -436,91 +436,180 @@ for decryption. To understand this algorithm, it is useful to define four
...
+compute overhead at load balancers, are discussed in
+{{distinguishing-attacks}}.
+
+The server concatenates the server ID and nonce into a single field, which is
+then split into equal halves. In successive passes, one of these halves is
+expanded into a 16B plaintext, encrypted with AES-ECB, and the result XORed with
+the other half. The diagram below shows the conceptual processing of a plaintext
+server ID and nonce into a connection ID. 'FO' stands for 'First Octet'.
+
+```
```
[the remark this answers is not in the document]
Comment (martinduke) [reply draft-ietf-quic-load-balancers.md:457]: I typed these commands but when I type 'aasvg' it says "command not found"

File: draft-ietf-quic-load-balancers.md [hunk 3]
```diff
@@ -436,91 +436,180 @@ for decryption. To understand this algorithm, it is useful to define four
...
+      |          \/                        |
+      |  +-+-+-+-+-+-+-+-+-+               |
+      |  |      left_2     |               |
+      |  +-+-+-+-+-+-+-+-+-+               |
+      |          |                         |
+      \/         \/                        \/
+   +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+   | FO  |            Ciphertext             |
+   +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+```
```
[the remark this answers is not in the document]
Comment (martinduke) [reply draft-ietf-quic-load-balancers.md:509]: OK, with these suggestions I was able to make a smaller diagram. Thanks!

File: draft-ietf-quic-load-balancers.md [hunk 4]
```diff
@@ -436,91 +436,180 @@ for decryption. To understand this algorithm, it is useful to define four
...
 ~~~pseudocode
-expand_right(0xaaba3c, 0x0b, 0x02) =
-                  0x0b020000000000000000000000aaba3c
+expand(0x06, 0x02, 0xaaba3c) = 0x0602aaba3c0000000000000000000000
 ~~~
 
-Similarly, truncate_left() and truncate_right() take the most significant and
-least significant bits, respectively, from a ciphertext. For example, to take
-28 bits of a ciphertext:
+Similarly, truncate(input, n) returns the n most significant octets of 'input'.
```
[the remark this answers is not in the document]
Comment (martinduke) [reply draft-ietf-quic-load-balancers.md:531]: Nope, I changed it from bits to octets and forgot to update there. Fixed now.

Good point about integers vs strings -- I got rid of the 'most significant' language with respect to octets.

**Comment by martinduke** [comment]
@martinthomson Thanks for all the formatting suggestions. It's cleaned up a bit, although I wasn't able to get aasvg to work. PTAL

**Comment by martinduke** [comment]
Correction. I got aasvg to work but it did not look good at all, probably due to user incompetence. If you would like to submit a PR after this that is prettier, I would welcome it.

File: draft-ietf-quic-load-balancers.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (martinthomson) [reply draft-ietf-quic-load-balancers.md]: You won't be able to run aasvg directly if you install it that way.  But the makefile should download it (assuming you have npm installed, that is) and run it for you.

File: draft-ietf-quic-load-balancers.md [hunk 5]
```diff
@@ -436,91 +436,158 @@ for decryption. To understand this algorithm, it is useful to define four
...
+      |    |          +-+-+-+-+-+      |
+      |    \/                          \/
+      |  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+      |  |      left_2     |      right_2    |
+      |  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+      |          |                  |
+      \/         \/                 \/
+   +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+   | FO  |            Ciphertext             |
+   +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
```
Comment (martinthomson) [inline draft-ietf-quic-load-balancers.md:488]:
```suggestion
~~~ aasvg
   +-----+-----------+-----------------------+
   | FO  | Server ID |         Nonce         |
   +--+--+-----------+-----+-----------------+
      |                    |
      |                    V
      |  +-----------------+-----------------+
      |  |      left_0     |      right_0    |
      |  +--+--------------+--------------+--+
      |     |                             |
      |     |                             |
      |     |         .--------.          V
      |     +-------->| AES-ECB +-------->⊕
      |     |         '--------'          |
      |     V             .--------.      | right_1
      |     ⊕<-----------+ AES-ECB |<-----+
      |     |             '--------'      |
      |     | left_1  .--------.          V
      |     +-------->| AES-ECB +-------->⊕
      |     |         '--------'          |
      |     V             .--------.      |
      |     ⊕<-----------+ AES-ECB |<-----+
      |     |             '--------'      |
      |     |                             |
      |     V                             V
      |  +-----------------+-----------------+
      |  |      left_2     |      right_2    |
      |  +-------+---------+--------+--------+
      |          |                  |
      V          V                  V
   +-----+-----------------------------------+
   | FO  |            Ciphertext             |
   +-----+-----------------------------------+
```

This looks something like this:

image

**Pull request merged**